datafusion/functions/src/datetime/extract.rs

```diff
@@ -0,0 +1,525 @@
+                IntervalUnit::Day => date_part(array.as_ref(), DatePart::Day)?,
+                IntervalUnit::Hour => date_part(array.as_ref(), DatePart::Hour)?,
+                IntervalUnit::Minute => date_part(array.as_ref(), DatePart::Minute)?,
+                IntervalUnit::Second => seconds_as_i32(array.as_ref(), Second)?,
+                IntervalUnit::Millisecond => seconds_as_i32(array.as_ref(), Millisecond)?,
+                IntervalUnit::Microsecond => seconds_as_i32(array.as_ref(), Microsecond)?,
+                IntervalUnit::Nanosecond => seconds_as_i32(array.as_ref(), Nanosecond)?,
+                // century and decade are not supported by `DatePart`, although they are supported in postgres
+                _ => return exec_err!("Date part '{part}' not supported"),
+            }
+        } else {
+            // special cases that can be extracted (in postgres) but are not interval units
+            match part_trim.to_lowercase().as_str() {
+                "qtr" | "quarter" => date_part(array.as_ref(), DatePart::Quarter)?,
+                "doy" => date_part(array.as_ref(), DatePart::DayOfYear)?,
+                "dow" => date_part(array.as_ref(), DatePart::DayOfWeekSunday0)?,
+                "isodow" => date_part(array.as_ref(), DatePart::DayOfWeekMonday0)?,
+                "epoch" => epoch(array.as_ref())?,
+                _ => return exec_err!("Date part '{part}' not supported"),
+            }
+        };
+
+        // Special adjustment for hour extraction on timezone-aware timestamps
+        if is_timezone_aware && part_trim.to_lowercase() == "hour" {
+            if let Some(tz_str) = &tz_str_opt {
+                let offset_hours = if tz_str.as_ref() == "+00:00" {
+                    0
+                } else {
+                    let sign = if tz_str.starts_with('+') { 1i32 } else { -1i32 };
+                    let hours_str = &tz_str[1..3];
+                    let hours: i32 = hours_str.parse().unwrap();
+                    sign * hours
+                };
+                let int_arr = as_int32_array(&arr)?;
+                let mut builder = PrimitiveBuilder::<Int32Type>::new();
+                for i in 0..arr.len() {
+                    if arr.is_null(i) {
+                        builder.append_null();
+                    } else {
+                        let v = int_arr.value(i);
+                        builder.append_value(v + offset_hours);
+                    }
+                }
+                arr = Arc::new(builder.finish());
+            }
+        }
+
+        Ok(if is_scalar {
+            ColumnarValue::Scalar(ScalarValue::try_from_array(arr.as_ref(), 0)?)
+        } else {
+            ColumnarValue::Array(arr)
+        })
+    }
+
+    fn aliases(&self) -> &[String] {
+        &[]
+    }
+
+    fn documentation(&self) -> Option<&Documentation> {
+        self.doc()
+    }
+}
+
+fn adjust_to_local_time<T: ArrowTimestampType>(ts: i64, tz: Tz) -> Result<i64> {
```

Comment:
Is this copied 1:1 from date_part.rs ?
There is yet another one at https://github.com/codetyri0n/datafusion/blob/5f2f5d7b79a3e8a773f591d4e27fc19227b4f226/datafusion/functions/src/datetime/to_local_time.rs#L296
Consider extracting this helper function in mod.rs (or a new file) and reusing it.